### DISCUSSION TITLE
NTDF/Blueprint Format

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#16656] [Feature requests]
I looked for a NTDF feature request or discussion but I didn't find the one I was thinking of. There was some thoughts about how data would be formatted. I recall the solution was ordained to be a list of tile elements with displacements from the center.

I want to get new track pieces rolling and one of the stopping blocks seems to be td6, so I want to get NTDF going again.

**Comment by duncanspumpkin** [comment]
Park files would be inappropriate for the use as a track design file but we will likely use some similar structure.

**Comment by spacek531** [comment]
I found the original issue for new track design format https://github.com/OpenRCT2/OpenRCT2/issues/8932

**Comment by spacek531** [comment]
I was talking with @X123M3-256 last night about adding new track elements before NTDF and he said "it wasn't a definite yes but they didn't immediately reject the idea."

Has this opinion changed? I'm very excited for new track elements.

**Comment by karst** [comment]
I do think the best integration for a NTDF would be something similar to the scenery manager plugin. 
TD6 not only has issues with new track pieces. It cannot save Trackitecture, addons like a transfer track (for either storage or for a tripple launch coaster or whatever) and there's a few bugs with not selecting the correct scenery and paths. 
You can save tracks that are completely non-functional too, because the break/boost speed isn't being taken into consideration with the testing, nor it's the transfer tracks are saved so that would make the coaster non functional aswell. 
My conclusion is that I don't think it'll be a big deal to add new track elements without a new save format for the tracks. 
A warning pop up would be nice tho, I suppose! Say the saving would detect a new element was used, it will tell that save will not be exported correctly. And if possible same with transfer tracks.

Lastly I would really love to see this being introduced as a feature aswell. The option looks like an awesome addition. https://github.com/oli414/AdvancedTrack/releases/

**Reply by spacek531** [reply]
If this is the multi-station plugin, I don't think NTDF/BF will need to support that from the outset. Trackitecture et al. is planned to be supported, from what I heard.

**Reply by karst** [reply]
I was thinking on the AdvancedTrack, making the functions savable, but have no interface without the plugin would be nice. Perhaps have some presets that are placed in the special elements list, so from a user perspective they're elements (like a switchback reverser, or a vertical flume lift). Keeping it userfriendly, but allow for more advanced things if wanted.

**Comment by ZehMatt** [comment]
One thing to note, I believe we decided to call the next iteration "Blueprints" instead of Track designs as the idea is to also allow scenery without rides.

**Comment by 733737** [comment]
would the extension be .blueprint or .track?

**Comment by 733737** [comment]
I think it might be an issue when sharing tracks if you don't know whether it is scenery like a building or a coaster, and you would have to completely redesign the prebuilt coaster placement window

**Comment by spacek531** [comment]
I had a brainwave for storing the locations of tile elements in uneconomic shapes without adding position data to the tile element. The key to this is that the "last for tile" flag is set properly. Store tile elements in columns so that the element after the last for tile-flagged element is a different column. It's probably easy to see where this is going at this point. Keep track of how many columns there are and have an array of position offsets that long. Reconstruct the ride by reading tile elements one by one and incrementing a counter after every last for tile-flagged element, and selecting the column position from the value at the counter.

**Comment by frutiemax** [comment]
I think TrackElements should be in their own file with a unique string id, and the track design would reference those unique string ids to indicate which track element to take.

**Comment by karst** [comment]
It might be a nice option to allow saving shoestringed rides as flat rides (rather than custom tracks). Perhaps it could only be saved as such using a plugin or a specific map similar to the track editor.

It would have to allow people to select the physical area of the ride, the rest would have no clearance and be invisible. It would also have to save the save-state which will be for opening and closing.
